In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB, MultinomialNB
from sklearn import tree

In [2]:
# load the SANK dataset:
sank_data=pd.read_csv("Sank.csv",header=0)

In [3]:
print(sank_data.shape)
print(list(sank_data.columns))

(891, 5)
['Alive', 'Class', 'Sex', 'Age', 'Fare']


In [4]:
sank_data

Alive  Class     Sex   Age     Fare
0        0      3    male  22.0   7.2500
1        1      1  female  38.0  71.2833
2        1      3  female  26.0   7.9250
3        1      1  female  35.0  53.1000
4        0      3    male  35.0   8.0500
..     ...    ...     ...   ...      ...
886      0      2    male  27.0  13.0000
887      1      1  female  19.0  30.0000
888      0      3  female   NaN  23.4500
889      1      1    male  26.0  30.0000
890      0      3    male  32.0   7.7500

[891 rows x 5 columns]

In [5]:
sank_data = sank_data.dropna()

In [6]:
sank_data

Alive  Class     Sex   Age     Fare
0        0      3    male  22.0   7.2500
1        1      1  female  38.0  71.2833
2        1      3  female  26.0   7.9250
3        1      1  female  35.0  53.1000
4        0      3    male  35.0   8.0500
..     ...    ...     ...   ...      ...
885      0      3  female  39.0  29.1250
886      0      2    male  27.0  13.0000
887      1      1  female  19.0  30.0000
889      1      1    male  26.0  30.0000
890      0      3    male  32.0   7.7500

[714 rows x 5 columns]

In [7]:
X = sank_data.loc[:, sank_data.columns != 'Alive']
y = sank_data.loc[:, sank_data.columns == 'Alive']
d = {'male': 1, 'female': 0}
X = X.replace({"Sex": d})

y = np.array(y)

In [8]:
X

Class  Sex   Age     Fare
0        3    1  22.0   7.2500
1        1    0  38.0  71.2833
2        3    0  26.0   7.9250
3        1    0  35.0  53.1000
4        3    1  35.0   8.0500
..     ...  ...   ...      ...
885      3    0  39.0  29.1250
886      2    1  27.0  13.0000
887      1    0  19.0  30.0000
889      1    1  26.0  30.0000
890      3    1  32.0   7.7500

[714 rows x 4 columns]

In [9]:
y

array([[0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [0],
    

In [10]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=0)

## SVM Model 

In [11]:
# SVM_Model = svm.SVC(kernel='linear')
SVM_Model = svm.SVC()
SVM_Model.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


SVC()

In [12]:
svm_parameters = {'kernel': ('linear', 'rbf'),'C':[1, 10, 20]}
SVM_GS_CV = GridSearchCV(SVM_Model, svm_parameters)
SVM_GS_CV.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return

GridSearchCV(estimator=SVC(),
             param_grid={'C': [1, 10, 20], 'kernel': ('linear', 'rbf')})

In [13]:
#Best model score
print("Best model score SVM Model :",SVM_GS_CV.score(X_train, y_train))
#Best Parameters
print("Best Parameters SVM Model :",SVM_GS_CV.best_params_)

Best model score SVM Model : 0.7758318739054291
Best Parameters SVM Model : {'C': 1, 'kernel': 'linear'}


## Logistic Regression Model

In [14]:
logreg = LogisticRegression(solver='liblinear',multi_class='auto')

# fit the model with data
logreg = logreg.fit(X_train,y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


In [15]:
LogRM_GS_parameters = {'C':[1, 5, 10]}
LogRM_GS_CV = GridSearchCV(logreg, LogRM_GS_parameters)
LogRM_GS_CV.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return

GridSearchCV(estimator=LogisticRegression(solver='liblinear'),
             param_grid={'C': [1, 5, 10]})

In [16]:
#Best model score Logistic Regression 
print("Best model score Logistic Regression :", LogRM_GS_CV.score(X_train, y_train))
#Best Parameters Logistic Regression 
print("Best Parameters Logistic Regression :",LogRM_GS_CV.best_params_)

Best model score Logistic Regression : 0.7863397548161121
Best Parameters Logistic Regression : {'C': 1}


## Random Forest Model

In [17]:
RandForC=RandomForestClassifier()
RandForC.fit(X_train,y_train)

/tmp/ipykernel_292/4222379738.py:2: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  RandForC.fit(X_train,y_train)


RandomForestClassifier()

In [18]:
RandForC_GS_parameters = {'n_estimators':[1, 5, 10]}
RandForC_GS_CV = GridSearchCV(RandForC, RandForC_GS_parameters)
RandForC_GS_CV.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:598: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:598: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:598: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  estimator.fit(X_train, y_train, **fit_params)
/opt/conda/lib/python3.9/site-packages/sklearn/model_selection/_validation.py:598: DataConversionWarning: A column-vector y was passed when a 

GridSearchCV(estimator=RandomForestClassifier(),
             param_grid={'n_estimators': [1, 5, 10]})

In [19]:
#Best model score Logistic Regression 
print("Best model score Random Forest :", RandForC_GS_CV.score(X_train, y_train))
#Best Parameters Logistic Regression 
print("Best Parameters Random Forest :",RandForC_GS_CV.best_params_)

Best model score Random Forest : 0.9737302977232924
Best Parameters Random Forest : {'n_estimators': 10}


# Naive_Bayes_Gaussian Model

In [20]:
GuassianN_Model = GaussianNB()
GuassianN_Model.fit(X_train,y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


GaussianNB()

In [21]:
GuassianN_GS_parameters = {}
GuassianN_GS_CV = GridSearchCV(GuassianN_Model, GuassianN_GS_parameters)
GuassianN_GS_CV.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return

GridSearchCV(estimator=GaussianNB(), param_grid={})

In [22]:
#Best model score Logistic Regression 
print("Best model score Naive_Bayes_Gaussian Model :", GuassianN_GS_CV.score(X_train, y_train))
#Best Parameters Logistic Regression 
print("Best Parameters Naive_Bayes_Gaussian Model :",GuassianN_GS_CV.best_params_)

Best model score Naive_Bayes_Gaussian Model : 0.7740805604203153
Best Parameters Naive_Bayes_Gaussian Model : {}


# Naive Bayes Multinomial Model

In [23]:
MultinomialN_Model = MultinomialNB()
MultinomialN_Model.fit(X_train,y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


MultinomialNB()

In [24]:
MultinomialN_parameters = {}
MultinomialN_GS_CV = GridSearchCV(MultinomialN_Model, MultinomialN_parameters)
MultinomialN_GS_CV.fit(X_train, y_train)

/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
/opt/conda/lib/python3.9/site-packages/sklearn/utils/validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return

GridSearchCV(estimator=MultinomialNB(), param_grid={})

In [25]:
#Best model score Logistic Regression 
print("Best model score Naive Bayes Multinomial Model :", MultinomialN_GS_CV.score(X_train, y_train))
#Best Parameters Logistic Regression 
print("Best Parameters Naive Bayes Multinomial Model :",MultinomialN_GS_CV.best_params_)

Best model score Naive Bayes Multinomial Model : 0.6900175131348512
Best Parameters Naive Bayes Multinomial Model : {}


# Decision Tree Model

In [26]:
dtree_Model = tree.DecisionTreeClassifier()
dtree_Model.fit(X_train,y_train)

DecisionTreeClassifier()

In [27]:
dtree_parameters = {'criterion':['gini','entropy']}
dtree_GS_CV = GridSearchCV(dtree_Model, dtree_parameters)
dtree_GS_CV.fit(X_train, y_train)

GridSearchCV(estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy']})

In [28]:
#Best model score Logistic Regression 
print("Best model score Decision Tree Model :", dtree_GS_CV.score(X_train, y_train))
#Best Parameters Logistic Regression 
print("Best Parameters Decision Tree Model :",dtree_GS_CV.best_params_)

Best model score Decision Tree Model : 0.9859894921190894
Best Parameters Decision Tree Model : {'criterion': 'entropy'}


### For the given DataSet Decision tree model suits best, because it's very fast compared to other models and with having score of around 99%. Random Forest algorithm seems to be performing very close to decision tree with best score of 97% whereas SVM seems to be performing worse with too much processing time and having best score of just around 78%. 